### ISSUE TITLE
NCCL version check needs to be done even if all symbols could not be loaded

### ISSUE THREAD

**Issue by stellaraccident** [iree-org/iree#13422]
We currently do this:

```
  // Resolve all symbols; this will fail if any required symbols are missing.
  if (iree_status_is_ok(status)) {
    status = iree_hal_cuda_nccl_dynamic_symbols_resolve_all(out_syms);
  }

  // Check the NCCL version compatibility
  int nccl_version = 0;
  if (iree_status_is_ok(status)) {
    status = NCCL_RESULT_TO_STATUS(out_syms, ncclGetVersion(&nccl_version));
  }
  if (iree_status_is_ok(status)) {
    int major = 0;
    int minor = 0;
    if (nccl_version < 20000) {
      major = nccl_version / 1000;
      minor = (nccl_version % 1000) / 100;
    } else {
      major = nccl_version / 10000;
      minor = (nccl_version % 10000) / 100;
    }
    if (nccl_version < NCCL_VERSION(NCCL_MAJOR, NCCL_MINOR, 0) ||
        major != NCCL_MAJOR) {
      status = iree_make_status(
          IREE_STATUS_INTERNAL,
          "NCCL version %d.%d found but at least version %d.%d is required",
          major, minor, NCCL_MAJOR, NCCL_MINOR);
    }
  }
```

Two things are wrong with this:

1. We fail with NOT_FOUND if symbols fail to load... which is the most likely thing that happens if we found a too old libnccl.so.
2. We return IREE_STATUS_INTERNAL instead of NOT_AVAILABLE on too old of a version.

We basically want to soft fail with UNAVAILABLE if a usable version is not found. This means that we should likely probe the version function first, run that check, then resolve all symbols.

This is hanging me up because in the course of some system update, I ended up with a libnccl.so 2.12 on my system, and we seem to be using some functions that appear to be from 2.17+.

**Comment by okkwon** [comment]
Thanks for reporting it. Here is a related PR under review: https://github.com/openxla/iree/pull/13403

Note that there is a bug in 2.17 so please use 2.15, which is tested.

**Comment by ScottTodd** [comment]
See also the discussion at https://github.com/openxla/iree/discussions/13400

**Comment by allieculp** [comment]
@okkwon Setting to 'in progress' - please update as needed.

**Comment by okkwon** [comment]
Resolved by #13432. Closing.
